File: jscomp/test/format_test.js
'use strict';

var Mt = require("./mt.js");
var List = require("../../lib/js/list.js");
var Bytes = require("../../lib/js/bytes.js");
var Pervasives = require("../../lib/js/pervasives.js");
var Caml_format = require("../../lib/js/caml_format.js");

var suites = {
  contents: /* [] */0
};

var test_id = {
  contents: 0
};

function eq(loc, x, y) {
  test_id.contents = test_id.contents + 1 | 0;
  suites.contents = {
    hd: [
      loc + (" id " + String(test_id.contents)),
      (function (param) {
          return {
                  TAG: /* Eq */0,
                  _0: x,
                  _1: y
                };
        })
    ],
    tl: suites.contents
  };
  
}

function eq3(loc, a, b, c) {
  eq(loc, a, b);
  eq(loc, b, c);
  return eq(loc, a, c);
}

var M = {};

eq("File \"format_test.ml\", line 28, characters 5-12", 7.875, 7.875);

eq("File \"format_test.ml\", line 31, characters 5-12", -7.875, -7.875);

eq3("File \"format_test.ml\", line 35, characters 6-13", Infinity, Number.POSITIVE_INFINITY, Pervasives.infinity);

eq3("File \"format_test.ml\", line 36, characters 6-13", -Infinity, Number.NEGATIVE_INFINITY, Pervasives.neg_infinity);

eq3("File \"format_test.ml\", line 37, characters 6-13", Pervasives.max_float, 1.79769313486231571e+308, Number.MAX_VALUE);

eq("File \"format_test.ml\", line 38, characters 5-12", Pervasives.classify_float(Infinity), /* FP_infinite */3);

eq("File \"format_test.ml\", line 39, characters 5-12", Pervasives.classify_float(Infinity), /* FP_infinite */3);

eq("File \"format_test.ml\", line 42, characters 5-12", Pervasives.min_float, 2.22507385850720138e-308);

eq("File \"format_test.ml\", line 43, characters 5-12", Pervasives.epsilon_float, 2.22044604925031308e-16);

eq("File \"format_test.ml\", line 44, characters 5-12", 4.94065645841e-324, 5e-324);

eq("File \"format_test.ml\", line 45, characters 5-12", 1.00000000000000022 - 1, Pervasives.epsilon_float);

eq("File \"format_test.ml\", line 47, characters 5-12", 1.11253692925360069e-308 / 2.22507385850720138e-308, 0.5);

eq("File \"format_test.ml\", line 49, characters 5-12", Pervasives.classify_float(1.11253692925360069e-308), /* FP_subnormal */1);

eq("File \"format_test.ml\", line 50, characters 5-12", 1.11253692925360069e-308, 1.11253692925360069e-308);

eq("File \"format_test.ml\", line 52, characters 5-12", 2.22507385850720138e-308, 2.22507385850720138e-308);

eq("File \"format_test.ml\", line 56, characters 5-12", (1 + 255 / 256) * 8, 15.96875);

eq("File \"format_test.ml\", line 59, characters 5-12", (1 + 4095 / 4096) * 8, 15.998046875);

eq("File \"format_test.ml\", line 62, characters 5-12", (1 + 65535 / 65536) * 8, 15.9998779296875);

function f(loc, ls) {
  return List.iter((function (param) {
                return eq(loc, Caml_format.float_of_string(param[0]), param[1]);
              }), ls);
}

f("File \"format_test.ml\", line 75, characters 6-13", {
      hd: [
        "0x3.fp+1",
        7.875
      ],
      tl: {
        hd: [
          " 0x3.fp2",
          15.75
        ],
        tl: {
          hd: [
            " 0x4.fp2",
            19.75
          ],
          tl: /* [] */0
        }
      }
    });

function sl(f) {
  return Caml_format.hexstring_of_float(f, -1, /* '-' */45);
}

function aux_list(loc, ls) {
  return List.iter((function (param) {
                return eq(loc, Caml_format.hexstring_of_float(param[0], -1, /* '-' */45), param[1]);
              }), ls);
}

var literals_0 = [
  7.875,
  "0x1.f8p+2"
];

var literals_1 = {
  hd: [
    0.3,
    "0x1.3333333333333p-2"
  ],
  tl: {
    hd: [
      Pervasives.infinity,
      "infinity"
    ],
    tl: {
      hd: [
        0.4,
        "0x1.999999999999ap-2"
      ],
      tl: {
        hd: [
          0.5,
          "0x1p-1"
        ],
        tl: {
          hd: [
            0.6,
            "0x1.3333333333333p-1"
          ],
          tl: {
            hd: [
              0.7,
              "0x1.6666666666666p-1"
            ],
            tl: {
              hd: [
                0.8,
                "0x1.999999999999ap-1"
              ],
              tl: {
                hd: [
                  0.9,
                  "0x1.ccccccccccccdp-1"
                ],
                tl: /* [] */0
              }
            }
          }
        }
      }
    }
  }
};

var literals = {
  hd: literals_0,
  tl: literals_1
};

aux_list("File \"format_test.ml\", line 109, characters 11-18", literals);

var s = Caml_format.hexstring_of_float(7.875, -1, /* '-' */45);

eq("File \"format_test.ml\", line 112, characters 5-12", Bytes.unsafe_to_string(Bytes.uppercase_ascii(Bytes.unsafe_of_string(s))), "0X1.F8P+2");

function scan_float(loc, s, expect) {
  return eq(loc, Caml_format.float_of_string(s), expect);
}

scan_float("File \"format_test.ml\", line 119, characters 13-20", "0x3f.p1", 126);

scan_float("File \"format_test.ml\", line 120, characters 13-20", "0x1.3333333333333p-2", 0.3);

List.iter((function (param) {
        return scan_float("File \"format_test.ml\", line 122, characters 13-20", param[1], param[0]);
      }), literals);

Mt.from_pair_suites("Format_test", suites.contents);

exports.suites = suites;
exports.test_id = test_id;
exports.eq = eq;
exports.eq3 = eq3;
exports.M = M;
exports.f = f;
exports.sl = sl;
exports.aux_list = aux_list;
exports.literals = literals;
exports.scan_float = scan_float;
/*  Not a pure module */

File: jscomp/test/format_test.ml
let suites :  Mt.pair_suites ref  = ref []
let test_id = ref 0
let eq loc x y = 
  incr test_id ; 
  suites := 
    (loc ^" id " ^ (string_of_int !test_id), (fun _ -> Mt.Eq(x,y))) :: !suites

let eq3 loc a b c = 
  eq loc a b ; 
  eq loc b c ; 
  eq loc a c 

let u () = "xx %s" ^^ "yy"

module M = struct 
  external infinity : float = "POSITIVE_INFINITY" 
  [@@bs.val]  [@@bs.scope "Number"]
  external neg_infinity : float = "NEGATIVE_INFINITY"
  [@@bs.val]  [@@bs.scope "Number"]
  external nan : float = "NaN"
  [@@bs.val]  [@@bs.scope "Number"]
  external max_float : float = "MAX_VALUE"
  [@@bs.val]  [@@bs.scope "Number"]
end 
let () = 
  eq __LOC__ (Format.asprintf (u ()) "x") ("xx x" ^ "yy")
  

#if OCAML_VERSION =~ ">4.03.0" then
let () = 
  eq __LOC__ (0x3.fp+1) (7.875);


  eq __LOC__ (-0x3.fp+1) (-7.875);  
  (* in standard, it is still infinity, but 
    ideally should give an warning
  *)
  eq3 __LOC__ 0x1p2047 M.infinity infinity;
  eq3 __LOC__ (-0x1p2047) M.neg_infinity neg_infinity;  
  eq3 __LOC__ max_float 0x1.ffff_ffff_ffff_fp+1023 M.max_float;  
  eq __LOC__ (classify_float 0x1.2p2047) FP_infinite;
  eq __LOC__ (classify_float 0x1.1p2047) FP_infinite;
  
  
  eq __LOC__ min_float 0x1p-1022;
  eq __LOC__ epsilon_float 0x1p-52;
  eq __LOC__ 0x0.0000_0000_0000_1p-1022 5e-324;
  eq __LOC__  
    (0x1.0000_0000_0000_1 -. 1.) epsilon_float;
  eq __LOC__
    (0x1p-1023 /. 0x1p-1022) 0x1p-1;
  eq __LOC__ (classify_float 0x1p-1023) FP_subnormal;
  eq __LOC__  
    0x1p-1023 0x0.8p-1022;
  eq __LOC__  
    0x0.ffff_ffff_ffff_ffff_ffp-1022
    0x1p-1022;

  eq __LOC__  
    ((1. +. 0xffp0 /. 0x100p0 ) *. 8.)
    (0x1.ffp3);
  eq __LOC__  
    ((1. +. 0xfffp0 /. 0x1000p0) *. 8.)
    0x1.fffp3;
  eq __LOC__  
    ((1. +. 0xffffp0 /. 0x10000p0) *. 8.)
    0x1.ffffp3    
  ;;
#end  
(*TODO: add scanf example *)

let f loc ls  = 
  List.iter  (fun (a,b) -> 
  eq loc (float_of_string a) b ) ls

#if OCAML_VERSION=~ ">4.03.0" then
let () = 
    f __LOC__ [
      "0x3.fp+1",  0x3.fp+1 ;
      " 0x3.fp2", 0x3.fp2;
      " 0x4.fp2", 0x4.fp2
      ];
    
#end     
;;


#if OCAML_VERSION =~ ">4.03.0" then 
let sl f = 
  Printf.sprintf "%h" f 

let aux_list loc ls =   
  List.iter (fun (a,b) -> 
    eq loc (sl a ) b 
  ) ls 
let literals = 
  [
    0x3.fp+1, "0x1.f8p+2" (* (1. +. 0xf8p0 /. 0x1_00p0) *. 4.*);
    0.3, "0x1.3333333333333p-2";
    infinity, "infinity";
    0.4, "0x1.999999999999ap-2";
    0.5, "0x1p-1";
    0.6, "0x1.3333333333333p-1";
    0.7, "0x1.6666666666666p-1";
    0.8, "0x1.999999999999ap-1";
    0.9, "0x1.ccccccccccccdp-1";
  ]

let () =   
  aux_list __LOC__ literals

let () = 
  eq __LOC__ (Printf.sprintf "%H" 0x3.fp+1) "0X1.F8P+2"  
let scan_float loc s expect = 
    Scanf.sscanf s "%h" (fun result -> eq loc result expect)

let () =     
  scan_float __LOC__ "0x3f.p1" 0x3f.p1;
  scan_float __LOC__ "0x1.3333333333333p-2" 0.3;
  List.iter (fun (a,b) -> 
  scan_float __LOC__ b a 
  ) literals
#end


#if 
  (* OCAML_VERSION =~ ">4.03.0" *) 0
then




#end
let () = Mt.from_pair_suites __MODULE__ !suites
